### ISSUE TITLE
Improving communication and message passing between components

### ISSUE THREAD

**Issue by sporto** [canjs/canjs#1209]
[VOTE HERE](http://bithub.com/event/43589)

# 1. Introduction

This is not really an issue but a document for discussion. Seems like putting it here is the best place for it.

The intention of this document is to explore patterns of communication between can components and discuss possible improvements and best practices.

It would be nice to aim for:

- Making common scenarios easier
- Establishing best practices and canonical patterns
- Having as little surprise as possible, it should be clear from looking at the code exactly what the intention is
- Aligning with future standards, specifically try to align with web components if possible
- Avoiding tight coupling between components, they should know as little as possible about their collaborators
- Making components easy to test, trying to keep the scope oblivious of the DOM

# 2. Communication from parent to children

## 2.1. Use case: We want to change the state on a sub component

This is currently supported by using view binding e.g.

```
<client-editor client="{client}" />
```

## 2.2. Use case: We want to send a message to a child component

Binding works well and it is easy to understand, but there are many situation where we have nothing to pass to the subcomponent. For example we just want to tell our sub-compontent to do something e.g. refresh.

### 2.2.1 Using the sub-component scope

It is possible to get a reference to the scope on a child component and call a method on that.

In sub-component:

```js
  scope: {
    refreshCount: 1,
    refresh: function (val) {
      var prev = this.attr('refreshCount');
      this.attr('refreshCount', prev + val);
    }
  },
```

In parent component:

```js
events: {
    '._btn_refresh_clients click': function () {
  	  var $comp = $('clients-chooser', this.element);
      $comp.scope().refresh(2);
    }
  }
```

[http://jsbin.com/daqox/1/edit]

This seems like a clean way to do this.

### 2.2.2. DOM methods

Another possible way is by attaching methods to the sub-component DOM.

In the sub-component:

```js
// when the component is inserted
// we attache a method to its DOM element
 events: {
    inserted: function ($ele) {
      var comp = this;
      $ele.get(0).refresh = function (args) {
       	// ... do something with comp
      }
    }
  }
```

In the parent:

```js
// when we want to send a message to the child component
// we find its element and then call the attached method
events: {
    '._btn_refresh click': function () {
  	  	var $comp = $('clients-chooser', this.element);
      	$comp.get(0).refresh(args);
    }
  }
```
  
[http://jsbin.com/hodosa/2/edit]

I have been using this method, but now I am not sure about the benefit of doing this in contrast to simply fetch the scope and call a method there.
  
### Possible improvements
  
It would be nice to have an streamlined way to create these methods on the DOM elements e.g.
  
```js
  can.Component.extend({
  	scope: {},
  	events: {},
  	domMethods: {
  		refresh: function (args) {
  			// ...
  		}
  	}
  });
```

Also, at the moment it is possible to get a reference to the scope on a component e.g.

```js
 $('child-component').scope();
```

What if we could get a reference to the component object and just call a method on that?

```js
 $('child-component').component().refresh(arg);
```

This would me unecessary the need to attach methods to the DOM.

## 2.3. Use case: We want to broadcast a message to all children

Calling a method on a child components is nice we have exactly one sub component we want to target. But what if there are many or they are nested? One possible way is to be able to broadcast an event to all sub elements.

Not sure if this is possible at the moment.

# 3. Communication from child to parent

## 3.1. Use case: We want to change the state on the parent controller

E.g We have a datepicker sub-component. 

Using model bindings is perfect for this. We can simply pass the model we want to change and allow the sub-component to change it directly.

```
<date-picker data="{booking.startsAt}" />

<date-picker data="{startsAt}" />
```

## 3.2. Use case: We want to send a message back to the parent

Again, there are many times when we don't have a model to bind, we just want to get a result back from the sub-component.

### 3.2.1. Triggering DOM events

Simply trigger an event an let it bubble it up the DOM, the parent controller should catch it. This is a well understood way of doing things. e.g.

In children component:

```js
 events: {
    'a click': function ($el, event) {
      var val = $el.html();
      // we simply trigger a DOM event with the value we want
      this.element.trigger('onSelected', val);
      return false;
    }
  }
```

In parent component:

```js
   events: {
    'clients-list onSelected': function (ele, event, val) {
	    // we receive the event using a normal listener
	    // and do something with it
     	this.scope.attr('choosen', val);
    }
  }
```

[http://jsbin.com/nuzew/1/edit]

There are a couple of issues with this approach:

- It is easy to introduce bugs if we change the name of the events in one place and not in the other. Is it possible to use constants?
- Events bubble all the way up, so potentially being catched by more things that what we want. We have to be careful with the naming of events or stop propagation.

#### Possible improvements

Web componets has a build in method in the components for something like this i.e. `fire`.

It would be nice to consider adding a covenience way of doing this. So instead of:

```
this.element.trigger('selected', val);
```

We could have:

```
this.fire('selected', val); // or emit or trigger
// this is the component object
```

### 3.2.2. Using a callback

A very well understood pattern in JS is to pass callbacks to objects, broadly used in jquery and jquery plugins. It would be nice to support something like this in can component.

e.g.

```
<clients-list on-selected="{onSelectedCallback}" />
```

At the moment this doesn't work because can cannot distinguish what is the intention here, e.g. to pass a callback or resolve the function.

A possible solution would be to introduce a stache convention for this:

```
<clients-list on-selected="{*onChosen}" />
```

### 3.2.3. Using can-event

Another possible way is to use can-event at the moment to replicate the callback approach.

```
<clients-list can-selected="onChosen" />
```

This looks a lot like the callback example above but note the `can-` prefix.

In the child component:

```js
  events: {
    'a click': function ($ele, event) {
      var val = $ele.html();
	      // ATM crafting an event object is necessary
			$(this.element).trigger({
            type: 'selected',
            val: val
          });
          // BUT it would be nice to simply do this:
          $(this.element).trigger('selected', val);
    }
  }
```

In the parent component:

```js
scope: {
    chosen: null,
    onChosen: function(context, elm, event) {
	    // ATM we need to extra the argument from the event object
      this.attr('chosen', event.val );
    }
  }

// BUT it would be nice to simply do this:

scope: {
    chosen: null,
    onChosen: function(context, elm, event, val) {
      this.attr('chosen', val);
    }
  }
```

[http://jsbin.com/koxobamo/22/edit]

### 3.2.4. Calling a method on the parent scope

Note that it is possible to call a method directly on the parent scope. But this should be avoided because of the tight coupling it creates.

## 3.3. Use case: We want to broadcast an event to the outside world

E.g. a child component broadcast that the current user name has been changed.

This is well supported by Triggering DOM events as explained above.

# 4. Communication between sibling components

## 4.1. Use case: We want to change the state on a sibling component

This is well supported by using model bindings, i.e. binding the same model in both components:

```js
<booking-editor>
	<client-editor client="{client}" />
	<client-finder current-client="{client}" />
</booking-editor>
```

## 4.2. Use case: We want to pass events between sibling components without model bindings

For example one component changes a value, we know that the state on the other components will be invalid because of this change, so we want to tell all these components to refresh themselves.

Here is a possible way of doing this:

- Trigger a DOM event from the initiating component to bubble up the DOM
- All the components interested on this events would listen to it by using `{document} nameOfEvent`

Example: [http://jsbin.com/rihid/1/edit]

**Comment by justinbmeyer** [comment]
With the new, much more flexible view-bindings, I think most of this is solved in 2.3.  The only lingering difficulty is in setting up parent knows about all children.  For now, you can do it like:

```
<parent>
  <child ($inserted)="addChild(%viewModel)"/>
  <child ($inserted)="addChild(%viewModel)"/>
</parent>
```

A bit of boilerplate, but works pretty well.

**Comment by pYr0x** [comment]
@justinbmeyer can we create a cheat sheet about communication between components? This would help a lot. Where can we create such a cheat sheet?
